x-pack/plugins/fleet/server/routes/data_streams/handlers.ts

```diff
@@ -4,157 +4,183 @@
+      };
+      managed_by?: string;
+      managed?: boolean;
+      [key: string]: any;
+    };
+    status: string;
+    template: string;
+    ilm_policy: string;
+    hidden: boolean;
+  }>;
+}
+
+interface ESDataStreamStatsResponse {
+  data_streams: Array<{
+    data_stream: string;
+    backing_indices: number;
+    store_size_bytes: number;
+    maximum_timestamp: number;
+  }>;
+}
+
 export const getListHandler: RequestHandler = async (context, request, response) => {
   const callCluster = context.core.elasticsearch.legacy.client.callAsCurrentUser;
+  const body: GetDataStreamsResponse = {
+    data_streams: [],
+  };
 
   try {
-    // Get stats (size on disk) of all potentially matching indices
-    const { indices: indexStats } = await callCluster('indices.stats', {
-      index: DATA_STREAM_INDEX_PATTERN,
-      metric: ['store'],
-    });
+    // Get matching data streams
+    const { data_streams: dataStreamsInfo } = (await callCluster('transport.request', {
+      method: 'GET',
+      path: `/_data_stream/${DATA_STREAM_INDEX_PATTERN}`,
+    })) as ESDataStreamInfoResponse;
+    const dataStreamsInfoByName = keyBy(dataStreamsInfo, 'name');
+
+    // Get data stream stats
+    const { data_streams: dataStreamStats } = (await callCluster('transport.request', {
+      method: 'GET',
+      path: `/_data_stream/${DATA_STREAM_INDEX_PATTERN}/_stats`,
+    })) as ESDataStreamStatsResponse;
+    const dataStreamsStatsByName = keyBy(dataStreamStats, 'data_stream');
+
+    // Combine data stream info
+    const dataStreams = merge(dataStreamsInfoByName, dataStreamsStatsByName);
+    const dataStreamNames = keys(dataStreams);
+
+    // Get all package SOs
+    const packageSavedObjects = await getPackageSavedObjects(context.core.savedObjects.client);
+    const packageSavedObjectsByName = keyBy(packageSavedObjects.saved_objects, 'id');
+    const packageMetadata: any = {};
+
+    // Query additional information for each data stream
+    const dataStreamPromises = dataStreamNames.map(async (dataStreamName) => {
+      const dataStream = dataStreams[dataStreamName];
+      const dataStreamResponse: DataStream = {
+        index: dataStreamName,
+        dataset: '',
+        namespace: '',
+        type: '',
+        package: dataStream._meta?.package?.name || '',
+        package_version: '',
+        last_activity_ms: dataStream.maximum_timestamp,
+        size_in_bytes: dataStream.store_size_bytes,
+        dashboards: [],
+      };
 
-    // Get all matching indices and info about each
-    // This returns the top 100,000 indices (as buckets) by last activity
-    const { aggregations } = await callCluster('search', {
-      index: DATA_STREAM_INDEX_PATTERN,
-      body: {
-        size: 0,
-        query: {
-          bool: {
-            must: [
-              {
-                exists: {
-                  field: 'data_stream.namespace',
+      // Query backing indices to extract data stream dataset, namespace, and type values
```

Comment:
I wonder if we could find a better way here for doing this. Instead of checking all the indices for it we could check what the values are in the mapping of the write index? Or if we run the query, could we ensure it only touches one index (most recent one) and not all of them? But then, what if this index was rollover but has no data inside yet?

Reply:
> what if this index was rollover but has no data inside yet

At some point I wrote the index name to be the last backing index (write index), but changed it for this exact reason :) I couldn't think of a better way to get these values but am open to suggestions.

Reply:
I would tend to trust ES here to be efficient in its aggregations, and leave it like this, until we know that it's a problem. Probing several indices until we find the newest one that has data is probably slower. Do we have a system with lots of rolled-over backing indices we could test with?